ext/drx/drx.c

```diff
@@ -2836,8 +2929,251 @@ static bool
+                                  params);
+                    break;
+                }
+            }
+        }
+        /* Intentionally not else if */
+        skip_unknown_instr_inc(DRX_DETECT_RESTORE_AVX512_SCATTER_EVENT_STATE_0, params);
+        break;
+    case DRX_DETECT_RESTORE_AVX512_SCATTER_EVENT_STATE_2:
+        if (instr_get_opcode(&params->inst) == OP_vextracti32x4) {
+            opnd_t dst0 = instr_get_dst(&params->inst, 0);
+            if (opnd_is_reg(dst0)) {
+                reg_id_t tmp_reg = opnd_get_reg(dst0);
+                if (!reg_is_strictly_xmm(tmp_reg))
+                    break;
+                params->the_scratch_xmm = tmp_reg;
+                advance_state(DRX_DETECT_RESTORE_AVX512_SCATTER_EVENT_STATE_3, params);
+                break;
+            }
+        }
+        /* Intentionally not else if */
+        skip_unknown_instr_inc(DRX_DETECT_RESTORE_AVX512_SCATTER_EVENT_STATE_0, params);
+        break;
+    case DRX_DETECT_RESTORE_AVX512_SCATTER_EVENT_STATE_3:
+        ASSERT(params->the_scratch_xmm != DR_REG_NULL,
+               "internal error: expected xmm register to be recorded in state "
+               "machine.");
+        if ((params->sg_info->scalar_value_size == OPSZ_4 &&
+             instr_get_opcode(&params->inst) == OP_vpextrd) ||
+            (params->sg_info->scalar_value_size == OPSZ_8 &&
+             instr_get_opcode(&params->inst) == OP_vpextrq)) {
+            ASSERT(opnd_is_reg(instr_get_src(&params->inst, 0)),
+                   "internal error: unexpected instruction format");
+            reg_id_t tmp_reg = opnd_get_reg(instr_get_src(&params->inst, 0));
+            if (tmp_reg == params->the_scratch_xmm) {
+                opnd_t dst0 = instr_get_dst(&params->inst, 0);
+                if (opnd_is_reg(dst0) && reg_is_gpr(opnd_get_reg(dst0))) {
+                    params->the_scratch_xmm = DR_REG_NULL;
+                    params->gpr_scratch_value = opnd_get_reg(dst0);
+                    advance_state(DRX_DETECT_RESTORE_AVX512_SCATTER_EVENT_STATE_4,
+                                  params);
+                    break;
+                }
+            }
+        }
+        skip_unknown_instr_inc(DRX_DETECT_RESTORE_AVX512_SCATTER_EVENT_STATE_0, params);
+        break;
+    case DRX_DETECT_RESTORE_AVX512_SCATTER_EVENT_STATE_4: {
+        if (!instr_is_reg_spill_or_restore(drcontext, &params->inst, NULL, NULL, NULL,
+                                           NULL)) {
+            if (instr_writes_memory(&params->inst)) {
+                opnd_t dst0 = instr_get_dst(&params->inst, 0);
+                if (opnd_is_memory_reference(dst0)) {
+                    opnd_t src0 = instr_get_src(&params->inst, 0);
+                    if (opnd_is_reg(src0)) {
+                        if (opnd_uses_reg(src0, params->gpr_scratch_value)) {
+                            if (opnd_uses_reg(dst0, params->gpr_scratch_index)) {
```

Comment:
Combine these 3 if's into a single if with && to make it easier to read and avoid all this gratuitous nesting.

Reply:
Done.